server/enterprise/elasticsearch/elasticsearch/elasticsearch.go

```diff
@@ -748,6 +754,49 @@ func (es *ElasticsearchInterfaceImpl) IndexChannel(rctx request.CTX, channel *mo
 	return nil
 }
 
+func (es *ElasticsearchInterfaceImpl) SyncBulkIndexChannels(rctx request.CTX, channels []*model.Channel, getUserIDsForChannel func(channel *model.Channel) ([]string, error), teamMemberIDs []string) *model.AppError {
+	if len(channels) == 0 {
+		return nil
+	}
+
+	es.mutex.RLock()
+	defer es.mutex.RUnlock()
+
+	if atomic.LoadInt32(&es.ready) == 0 {
+		return model.NewAppError("Elasticsearch.SyncBulkIndexChannels", "ent.elasticsearch.not_started.error", map[string]any{"Backend": model.ElasticsearchSettingsESBackend}, "", http.StatusInternalServerError)
+	}
+
+	indexName := *es.Platform.Config().ElasticsearchSettings.IndexPrefix + common.IndexBaseChannels
+	metrics := es.Platform.Metrics()
+
+	for _, channel := range channels {
+		userIDs, err := getUserIDsForChannel(channel)
+		if err != nil {
+			return model.NewAppError("Elasticsearch.SyncBulkIndexChannels", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err)
+		}
+
+		searchChannel := common.ESChannelFromChannel(channel, userIDs, teamMemberIDs)
+
+		err = es.syncBulkProcessor.IndexOp(types.IndexOperation{
+			Index_: model.NewPointer(indexName),
+			Id_:    model.NewPointer(searchChannel.Id),
+		}, searchChannel)
+		if err != nil {
+			return model.NewAppError("Elasticsearch.SyncBulkIndexChannels", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err)
+		}
+
+		if metrics != nil {
+			metrics.IncrementChannelIndexCounter()
+		}
+	}
+
+	if err := es.syncBulkProcessor.flush(); err != nil {
```

Comment:
I guess the only slight problem with the syncProcessor is we are not flushing at all until the very end, which means that potentially a lot of items will be in memory until they are flushed. There's also another problem of very large request sizes on which there is a limit in ES.

Wondering if we can keep the periodicFlusher intact (which will ensure flushing when batch size exceeds a threshold), but also add a force flush at the end which should take care of any pending items in the buffer?

Reply:
> Maybe a simple solution would be to add another condition to the timer that flushes the data every, let's say, 10MiB. Thoughts?

That could certainly be a solution and probably what the old bulk indexer had. But even with the current solution, if `LiveIndexingBatchSize` is too high, we can always ask the customer to lower it which would make it flush earlier. Alternatively, I think tracking `r.buf.Len()` to check the size of the buffer and then flush it when it crosses 10MiB is also fine. It's not that much code.